### ISSUE TITLE
race condition in LoggerSilence

### ISSUE THREAD

**Issue by gucki** [rails/rails#22425]
We are using rails 4.2.4 with puma and noticed that sometimes the logging just stops. After a lot of debugging it turns this is caused by a race condition in activesupport/lib/active_support/logger_silence.rb.

    [
        [ 0] "/home/gucki/Coding/projects/test/backend/config/application.rb:67:in `level='",
        [ 1] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/logger_silence.rb:15:in `silence'",
        [ 2] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activerecord-session_store-0.1.1/lib/action_dispatch/session/active_record_store.rb:78:in `set_session'",
        [ 3] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:342:in `commit_session'",
        [ 4] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:226:in `context'",
        [ 5] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:220:in `call'",
        [ 6] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/cookies.rb:560:in `call'",
        [ 7] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activerecord-4.2.4/lib/active_record/query_cache.rb:36:in `call'",
        [ 8] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activerecord-4.2.4/lib/active_record/connection_adapters/abstract/connection_pool.rb:653:in `call'",
        [ 9] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'",
        [10] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/callbacks.rb:88:in `__run_callbacks__'",
        [11] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/callbacks.rb:778:in `_run_call_callbacks'",
        [12] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/callbacks.rb:81:in `run_callbacks'",
        [13] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/callbacks.rb:27:in `call'",
        [14] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/remote_ip.rb:78:in `call'",
        [15] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'",
        [16] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'",
        [17] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/rack/logger.rb:38:in `call_app'",
        [18] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/rack/logger.rb:20:in `block in call'",
        [19] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/tagged_logging.rb:68:in `block in tagged'",
        [20] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/tagged_logging.rb:26:in `tagged'",
        [21] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/tagged_logging.rb:68:in `tagged'",
        [22] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/rack/logger.rb:20:in `call'",
        [23] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/request_store-1.2.0/lib/request_store/middleware.rb:8:in `call'",
        [24] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/actionpack-4.2.4/lib/action_dispatch/middleware/request_id.rb:21:in `call'",
        [25] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/methodoverride.rb:22:in `call'",
        [26] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/runtime.rb:18:in `call'",
        [27] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/activesupport-4.2.4/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'",
        [28] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/rack-1.6.4/lib/rack/sendfile.rb:113:in `call'",
        [29] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/engine.rb:518:in `call'",
        [30] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/application.rb:165:in `call'",
        [31] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/railtie.rb:194:in `public_send'",
        [32] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/railties-4.2.4/lib/rails/railtie.rb:194:in `method_missing'",
        [33] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/puma-2.15.0/lib/puma/configuration.rb:79:in `call'",
        [34] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/puma-2.15.0/lib/puma/server.rb:541:in `handle_request'",
        [35] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/puma-2.15.0/lib/puma/server.rb:388:in `process_client'",
        [36] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/puma-2.15.0/lib/puma/server.rb:270:in `block in run'",
        [37] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/puma-2.15.0/lib/puma/thread_pool.rb:106:in `call'",
        [38] "/home/gucki/.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/gems/puma-2.15.0/lib/puma/thread_pool.rb:106:in `block in spawn_thread'"
    ]

The first line of the backtrace is just my debug hook. But as you can see each request is using the logger silencer.  

    def silence(temporary_level = Logger::ERROR)
      if silencer
        begin
          old_logger_level, self.level = level, temporary_level
          yield self
        ensure
          self.level = old_logger_level
        end
      else
        yield self
      end
    end

Now if two or more thread enter that method in parallel, there's a change that only the first call sees the real log level and a second call sees the temporaty_level, because the real log level was overriden by the first call but not reset yes. Now when this call ends the log level is not properly reset (it's reset to the wrong temporary value) and thus logging stays disabled.

We need a mutex here. Either a new one or reuse the one from the logger? Any suggestions?

**Comment by gucki** [comment]
Here's a monkey-patch:

    module SafeLoggerSilence
      extend ActiveSupport::Concern

      included do
        cattr_accessor :mutex
        self.mutex = Mutex.new
      end

      def silence(temporary_level = Logger::ERROR)
        mutex.synchronize do
          super(temporary_level)
        end
      end
    end

    class ActiveSupport::Logger
      include SafeLoggerSilence
    end

**Comment by arthurnn** [comment]
we need to use the mutex from the logger, so the other operations are in sync with this one too. We saw this problem before, and thats why I changed ruby to expose that mutex. see https://github.com/ruby/ruby/pull/541 
That was merged in ruby, I am still not sure which version will include that, but we will be able to use it after it is released.

**Comment by gucki** [comment]
@arthurnn Yeah, that would be the perfect fix. But will your ruby patch be backported to older rubies too?  Meanwhile we could fail back to some hacky mutex access using instance_variable_get?

**Comment by arthurnn** [comment]
Thats a good point, I think we could use the `@mutex` until we have that in a stable ruby version, lets just add a TODO above the `instance_variable_get` pointing to the ruby patch, so one day we can fully replace it.
Do you wanna work in a PR for this?

**Comment by matthewd** [comment]
This is a duplicate of #16885, #20490, and #20507.
